src/hardware/uni-t-ut8802e/api.c

```diff
@@ -0,0 +1,219 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <config.h>
+#include "protocol.h"
+
+static const uint32_t scanopts[] = {
+    SR_CONF_CONN,
+};
+
+static const uint32_t drvopts[] = {
+    SR_CONF_MULTIMETER,
+};
+
+static const uint32_t devopts[] = {
+    SR_CONF_CONTINUOUS,
+    SR_CONF_LIMIT_SAMPLES | SR_CONF_GET | SR_CONF_SET,
+    SR_CONF_LIMIT_MSEC | SR_CONF_GET | SR_CONF_SET,
+    SR_CONF_DATA_SOURCE | SR_CONF_GET | SR_CONF_LIST,
+};
+
+
+static const char *channel_names[] = {
+    "Main",
+};
+
+static const char *data_sources[] = {
+    "Live",
+};
+
+static struct sr_dev_driver uni_t_ut8802e_driver_info;
+
+static GSList *scan(struct sr_dev_driver *di, GSList *options)
+{
+    struct sr_dev_inst *sdi;
+    struct dev_context *devc;
+    struct sr_serial_dev_inst *serial;
+    struct sr_config *src;
+    const char *conn, *serialcomm;
+    char conn_id[64];
+    int rc;
+    size_t idx;
+    GSList *l ,*devices;
+
+    conn = NULL;
+    serialcomm = "9600/8n1";
+    for (l = options; l; l = l->next) {
+        src = l->data;
+        if (src->key == SR_CONF_CONN)
+            conn = g_variant_get_string(src->data, NULL);
+    }
+
+    if (!conn)
+        return NULL;
+
+    devices = NULL;
+    serial = sr_serial_dev_inst_new(conn, serialcomm);
+    rc = serial_open(serial, SERIAL_RDWR);
+    snprintf(conn_id, sizeof(conn_id), "%s", serial->port);
+    if (rc != SR_OK) {
+        serial_close(serial);
+        sr_serial_dev_inst_free(serial);
+        return devices;
+    }
+
+    sdi = g_malloc0(sizeof(*sdi));
+    sdi->status = SR_ST_INACTIVE;
+    sdi->vendor = g_strdup("UNI-T");
+    sdi->model = g_strdup("UT8802E");
+    sdi->inst_type = SR_INST_SERIAL;
+    sdi->conn = serial;
+    sdi->connection_id = g_strdup(conn_id);
+    devc = g_malloc0(sizeof(*devc));
+    sdi->priv = devc;
+    sr_sw_limits_init(&devc->limits);
+    for (idx = 0; idx < ARRAY_SIZE(channel_names); idx++) {
+        sr_channel_new(sdi, idx, SR_CHANNEL_ANALOG, TRUE,
+                channel_names[idx]);
+    }
+    devices = g_slist_append(devices, sdi);
+
+    return std_scan_complete(di, devices);
+
+}
+
+
+static int config_get(uint32_t key, GVariant **data,
+    const struct sr_dev_inst *sdi, const struct sr_channel_group *cg)
+{
+    struct dev_context *devc;
+
+    (void)cg;
+
+    devc = sdi->priv;
+
+    switch (key) {
+    case SR_CONF_CONN:
+        *data = g_variant_new_string(sdi->connection_id);
+        break;
+    case SR_CONF_LIMIT_FRAMES:
+    case SR_CONF_LIMIT_SAMPLES:
+    case SR_CONF_LIMIT_MSEC:
+        if (!devc)
+            return SR_ERR_ARG;
+        return sr_sw_limits_config_get(&devc->limits, key, data);
+    case SR_CONF_DATA_SOURCE:
+        *data = g_variant_new_string(data_sources[0]);
+        break;
+    // TO DO ADD SR_CONF_MEASURED_QUANTITY and SR_CONF_RANGE
```

Comment:
Please don't leave TODOs in the code that's going up to mainline. It's ok if your driver doesn't yet support all features the hardware has, but the driver should just be functional for the stuff it does support. You can make a note of features it still needs in the commit message or (much preferred) the wiki.